burn-tensor/src/tensor/api/base.rs

```diff
@@ -168,13 +173,103 @@ where
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// let data = Data::from([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]);
+    /// let tensor: burn_tensor::Tensor<TestBackend, 3, burn_tensor::Int> = Tensor::from_data(data);
+    /// let mut acc = String::new();
+    /// let mut multi_index = vec![0; 3];
+    /// tensor.display_recursive(&mut acc, 0, &mut multi_index);
+    /// assert_e!(acc, "[[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]");
+    /// ```
+    fn display_recursive(&self, acc: &mut String, depth: usize, multi_index: &mut [usize]) {
+        if depth == 0 {
+            acc.push('[');
+        }
+
+        if depth == self.dims().len() - 1 {
+            // if we are at the innermost dimension, just push its elements into the accumulator
+            for i in 0..self.dims()[depth] {
+                if i > 0 {
+                    acc.push_str(", ");
+                }
+                multi_index[depth] = i;
+                let range: [std::ops::Range<usize>; D] =
+                    core::array::from_fn(|i| multi_index[i].clone()..multi_index[i].clone() + 1);
+                let elem = &self.clone().index(range).to_data().value[0];
+                acc.push_str(&format!("{:?}", elem));
+            }
+        } else {
+            // otherwise, iterate through the current dimension and recursively display the inner tensors
+            for i in 0..self.dims()[depth] {
+                if i > 0 {
+                    acc.push_str(", ");
+                }
+                acc.push('[');
+                multi_index[depth] = i;
+                self.display_recursive(acc, depth + 1, multi_index);
+                acc.push(']');
+            }
+        }
+
+        if depth == 0 {
+            acc.push(']');
+        }
+    }
+}
+
+/// Pretty print tensors
+impl<B, const D: usize, K> std::fmt::Display for Tensor<B, D, K>
+where
+    B: Backend,
+    B::IntElem: std::fmt::Display,
+    K: BasicOps<B>,
+    <K as BasicOps<B>>::Elem: Debug,
+{
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Tensor {{")?;
+        write!(f, "  data: ")?;
+
+        let mut acc = String::new();
+        let mut multi_index = vec![0; D];
+        self.display_recursive(&mut acc, 0, &mut multi_index);
+        write!(f, "{}", acc)?;
+        writeln!(f, ",")?;
+        writeln!(f, "  shape:   {:?},", self.dims())?;
+        writeln!(f, "  device:  {:?},", self.device())?;
+        writeln!(f, "  backend: {:?},", B::name())?;
+        writeln!(f, "  dtype:   {:?},", K::elem_type())?;
+        write!(f, "}}")
+    }
+}
+
 /// Trait that list all operations that can be applied on all tensors.
 ///
 /// # Warnings
 ///
 /// This is an internal trait, use the public API provided by [tensor struct](Tensor).
 pub trait BasicOps<B: Backend>: TensorKind<B> {
-    type Elem;
+    type Elem: 'static;
```

Comment:
@nathanielsimard Will doing this impact anything? From my understanding, adding the `'static` means any type implementing `BasicOps` must have a static lifetime for its Elem type, which shouldn't be a problem since numeric types like i32, f32, etc., have a static lifetime..

Reply:
Yes this is fine!